File: sdk/go/failure.go
package shojiku

import "fmt"

// Step is the lifecycle step a [Failure] belongs to.
//
// Always one of these three — the SDK's OWN vocabulary. The engine's report
// carries a step of its own naming an INTERNAL stage (`render`, `validate`),
// and passing that through would make the trace's step mean different things
// depending on which layer refused. What the engine said specifically is the
// failure's [Failure.Kind].
type Step string

// The three lifecycle steps.
const (
	StepGenerate Step = "generate"
	StepSign     Step = "sign"
	StepVerify   Step = "verify"
)

// Failure says why a lifecycle operation did not produce what was asked for.
//
// A VALUE, not an error. The shape takes effect-ts's Cause as its conceptual
// reference: which step failed, what class of thing went wrong, and — when
// one failure happened because of another — the chain underneath it, all
// inspectable rather than unwound. No framework is involved; only the idea
// that a failure is data.
type Failure struct {
	step        Step
	kind        string
	message     string
	diagnostics []Diagnostic
	cause       *Failure
}

// Step is the lifecycle step, which is always this package's own.
func (f *Failure) Step() Step { return f.step }

// Kind is a stable machine-readable class. Engine-side kinds come straight
// off the report; host-side ones are this package's own (`template_name`,
// `template_escapes_root`, …).
func (f *Failure) Kind() string { return f.kind }

// Message is the prose, already bounded by whoever produced it.
func (f *Failure) Message() string { return f.message }

// Diagnostics are what the engine noticed, when this failure carries any.
func (f *Failure) Diagnostics() []Diagnostic { return f.diagnostics }

// Cause is the failure underneath this one, or nil.
func (f *Failure) Cause() *Failure { return f.cause }

// Causes is this failure and everything under it, outermost first. What you
// log when you want the whole story rather than only its headline.
func (f *Failure) Causes() []*Failure {
	chain := []*Failure{f}
	for cause := f.cause; cause != nil; cause = cause.cause {
		chain = append(chain, cause)
	}
	return chain
}

func (f *Failure) String() string {
	return fmt.Sprintf("%s/%s: %s", f.step, f.kind, f.message)
}

File: sdk/php/src/Failure.php
<?php

declare(strict_types=1);

namespace Shojiku;

/**
 * Why a lifecycle operation did not produce what was asked for.
 *
 * A VALUE, not an exception. The shape takes effect-ts's `Cause` as its
 * conceptual reference: which step failed, what class of thing went wrong,
 * and — when one failure happened because of another — the chain underneath
 * it, all inspectable rather than unwound. No effect framework is involved;
 * only the idea that a failure is data.
 */
final class Failure
{
    /**
     * @param list<Diagnostic> $diagnostics
     */
    public function __construct(
        private readonly Step $step,
        private readonly string $kind,
        private readonly string $message,
        private readonly array $diagnostics = [],
        private readonly ?self $cause = null,
    ) {
    }

    /**
     * The lifecycle step, which is always this package's own.
     *
     * The engine's report names an INTERNAL stage in its own `step` field
     * (`render`, `validate`); reading it here would make this field mean
     * different things depending on which layer refused. What the engine said
     * specifically is {@see self::kind()}.
     */
    public function step(): Step
    {
        return $this->step;
    }

    /**
     * A stable machine-readable class. Engine-side kinds come straight off
     * the report; host-side ones are this package's own (`template_name`,
     * `key_unreadable`, …).
     */
    public function kind(): string
    {
        return $this->kind;
    }

    public function message(): string
    {
        return $this->message;
    }

    /**
     * @return list<Diagnostic>
     */
    public function diagnostics(): array
    {
        return $this->diagnostics;
    }

    public function cause(): ?self
    {
        return $this->cause;
    }

    /**
     * This failure and everything under it, outermost first. What you log
     * when you want the whole story rather than only its headline.
     *
     * @return list<self>
     */
    public function causes(): array
    {
        $chain = [$this];
        $cause = $this->cause;
        while ($cause !== null) {
            $chain[] = $cause;
            $cause = $cause->cause;
        }

        return $chain;
    }

    public function __toString(): string
    {
        return sprintf('%s/%s: %s', $this->step->value, $this->kind, $this->message);
    }
}
